http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/CrtRequestExecutor.java

```diff
@@ -124,23 +193,138 @@ private void executeRequest(CrtRequestContext executionContext,
+                                      HttpHeader[] nextHeaders) {
+            if (blockType == HttpHeaderBlock.MAIN.getValue()) {
+                for (HttpHeader h : nextHeaders) {
+                    responseBuilder.appendHeader(h.getName(), h.getValue());
+                }
+            }
+
+            if (responseBuilder.statusCode() == 0) {
+                responseBuilder.statusCode(responseStatusCode);
+            }
+
+        }
+
+        @Override
+        public void onResponseComplete(HttpStream stream, int errorCode) {
+            // always close the connection on a 5XX response code.
+            if (HttpStatusFamily.of(responseBuilder.statusCode()) == HttpStatusFamily.SERVER_ERROR) {
+                crtConn.shutdown();
+            }
+
+            if (errorCode == 0) {
+                requestCompletionFuture.complete(responseBuilder.build());
+            } else {
+                HttpException rootException = new HttpException(errorCode);
+                Throwable toThrow = HttpClientConnection.isErrorRetryable(rootException) ? rootException :
+                                    new IOException(rootException);
+                requestCompletionFuture.completeExceptionally(toThrow);
+            }
+            stream.close();
+            crtConn.close();
+
+            if (simplePublisher != null) {
+                simplePublisher.complete();
+            }
+        }
+
+        @Override
+        public int onResponseBody(HttpStream stream, byte[] bodyBytesIn) {
+
+            // Per the client conformance tests, unless there's an actual body the response stream must be null.
+            // This is function is only called in one of the CRT's IO threads, or is otherwise sequentially ordered.
+            // It is volatile in the declaration for thread visibility reasons, NOT concurrency control.
+            if (inputStreamSubscriber == null && simplePublisher == null) {
+                inputStreamSubscriber = new InputStreamSubscriber();
+                simplePublisher = new SimplePublisher<>();
+                simplePublisher.subscribe(inputStreamSubscriber);
+                responseBuilder.content(AbortableInputStream.create(inputStreamSubscriber));
+            }
+
+            CompletableFuture<Void> writeFuture = simplePublisher.send(ByteBuffer.wrap(bodyBytesIn));
+
+            // go ahead and let back pressure know we consumed the data.
+            if (writeFuture.isDone() && !writeFuture.isCompletedExceptionally()) {
+                return bodyBytesIn.length;
+            }
+
+            writeFuture.whenComplete((result, failure) -> {
+                if (failure != null) {
+                    requestCompletionFuture.completeExceptionally(failure);
+                    return;
+                }
+
+                // otherwise, increment the window upon buffer consumption.
+                stream.incrementWindow(bodyBytesIn.length);
+            });
```

Comment:
We essentially do the same thing in two different ways here. Can we just keep the `whenComplete` call and remove 275-277 to avoid issues with these paths diverting in the future?

`whenComplete` should run synchronously anyway if the future is done at this point

Reply:
Thanks, fixed.